File: medium/kadanesAlgorithm.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int kadanesAlgorithm(const std::vector<int> &array);

int main() {
    std::vector<int> array{1, -6, 4, 3, -3, 8};

    std::cout << kadanesAlgorithm(array) << std::endl;
}

int kadanesAlgorithm(const std::vector<int> &array) {
    int accumulate{array[0]};
    int maxSum{accumulate};

    for (int i{1}; i < static_cast<int>(array.size()); ++i) {
        accumulate = std::max(accumulate + array[i], array[i]);
        maxSum = std::max(maxSum, accumulate);
    }

    return maxSum;
}

File: medium/kadanesAlgorithm.py
def kadanes_algorithm(array):
    accumulate = array[0]
    max_sum = accumulate

    for index in range(1, len(array)):
        accumulate = max(accumulate + array[index], array[index])
        max_sum = max(max_sum, accumulate)

    return max_sum


def main():
    array = [1, -6, 4, 3, -3, 8]
    print(kadanes_algorithm(array))


if __name__ == "__main__":
    main()
